### PR TITLE
#2772 - Don't recalculate command map on world change

### LINKED ISSUE #2588: (1.13.2) It takes 20-120ms to teleport a player to a different world.
Teleporting a player to a different world causes extra latancy within the server. This is not an issue of loading chunks because there is no latency teleporting a player to an unloaded area. 

I have had reported lag varying between 20-120 milliseconds and has no other trigger to cause it to go higher or lower.

I am running paper-648 (1.13.2)

### PR DESCRIPTION
**Pull request by perseus5**
Continuation of #2772, wanted to speed that along...
Fixes #2588

### CODE REVIEW

**Comment by perseus5** [comment]
Fixed the bug that left you in an unpleasant state.

**Review by BillyGalbreath — APPROVED** [review]

**Review by Puremin0rez — APPROVED** [review]
I can confirm that any issues I experienced initially are now resolved by the latest changes.

File: Spigot-Server-Patches/0437-Dont-recalculate-permissions-on-world-change.patch [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+@@ -653,4 +653,10 @@ public class PaperWorldConfig {
+         disableHopperMoveEvents = getBoolean("hopper.disable-move-event", disableHopperMoveEvents);
+         log("Hopper Move Item Events: " + (disableHopperMoveEvents ? "disabled" : "enabled"));
+     }
++    
++    public boolean recalculatePermsOnWorldChange = false;
++    private void recalculatePermsOnWorldChange() {
++    	recalculatePermsOnWorldChange = getBoolean("recalculate-perms-on-world-change", recalculatePermsOnWorldChange);
```
Comment (Spottedleaf) [inline Spigot-Server-Patches/0437-Dont-recalculate-permissions-on-world-change.patch]: this indentation

File: Spigot-Server-Patches/0437-Dont-recalculate-permissions-on-world-change.patch [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+index bce502181..cc7e029b5 100644
+--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+@@ -653,4 +653,10 @@ public class PaperWorldConfig {
+         disableHopperMoveEvents = getBoolean("hopper.disable-move-event", disableHopperMoveEvents);
+         log("Hopper Move Item Events: " + (disableHopperMoveEvents ? "disabled" : "enabled"));
+     }
++    
++	public boolean recalculatePermsOnWorldChange = false;
```
Comment (Spottedleaf) [inline Spigot-Server-Patches/0437-Dont-recalculate-permissions-on-world-change.patch]: no tabs

File: Spigot-Server-Patches/0437-Dont-recalculate-permissions-on-world-change.patch [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+index bce502181..c41927c80 100644
+--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+@@ -653,4 +653,10 @@ public class PaperWorldConfig {
+         disableHopperMoveEvents = getBoolean("hopper.disable-move-event", disableHopperMoveEvents);
+         log("Hopper Move Item Events: " + (disableHopperMoveEvents ? "disabled" : "enabled"));
+     }
++
++    public boolean recalculatePermsOnWorldChange = false;
```
Comment (electronicboy) [inline Spigot-Server-Patches/0437-Dont-recalculate-permissions-on-world-change.patch]: I'm iffy on this, this is one of those things which I believe should be opt-in, people use permissions across world, let's not break their setups

File: Spigot-Server-Patches/0437-Dont-recalculate-permissions-on-world-change.patch [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+@@ -653,4 +653,10 @@ public class PaperWorldConfig {
+         disableHopperMoveEvents = getBoolean("hopper.disable-move-event", disableHopperMoveEvents);
+         log("Hopper Move Item Events: " + (disableHopperMoveEvents ? "disabled" : "enabled"));
+     }
++
++    public boolean recalculatePermsOnWorldChange = true;
++    private void recalculatePermsOnWorldChange() {
++        recalculatePermsOnWorldChange = getBoolean("recalculate-perms-on-world-change", recalculatePermsOnWorldChange);
```
Comment (BillyGalbreath) [inline Spigot-Server-Patches/0437-Dont-recalculate-permissions-on-world-change.patch]: This should probably be in the `settings` section of the config

**Review by BillyGalbreath — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0437-Dont-recalculate-permissions-on-world-change.patch [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+index 1d7d1ffbf..a580611c8 100644
+--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
+@@ -425,4 +425,9 @@ public class PaperConfig {
+             */
+         }
+     }
++
++    public static boolean recalculatePermsOnWorldChange = true;
++    private void recalculatePermsOnWorldChange() {
```
Comment (Spottedleaf) [inline Spigot-Server-Patches/0437-Dont-recalculate-permissions-on-world-change.patch]: needs to be static

**Comment by SrBedrock** [comment]
Could this break permissions that depend on the world as a context?

Example: Player has a permission that is only given in the world world_nether.
Using luckperms would be something like:
/lp user `<player>` permission set `my.permission` true `world=world_nether`

If the player with that permission goes to the world called world_nether, would the permission have no effect?
This permission is verified when the player dies.

**Comment by Artuto** [comment]
This is only for the command list

**Comment by SrBedrock** [comment]
Thanks @Artuto

**Comment by Plutoren** [comment]
I am getting a 700ms lag spike on every player's world change and I'm told this could fix it so... bump

**Comment by clrxbl** [comment]
There is no need to "bump" PR's on Github. If you think this can fix it, build a Paper jar w/ the patch and see if it does.

**Comment by Plutoren** [comment]
Ahh, my bad, I only did so at the suggestion of someone on the discord, and it appears someone has already tested it and confirmed it fixes the issue. I can test it myself if that would help the pr get merged it just seems like a pretty critical issue (I have 10+ world changes per minute on my server which means my server is stalling about 15% of the time) But I am not trying to break the workflow around here I just talked about it on discord and someone said this branch was the cure and to bump it so I hope it gets merged one day :)

**Comment by electronicboy** [comment]
Sorry for the immense delay on this,
I'm confused about the total intent of this, the PR states that it's preventing the command structure from being recalculated, but, it doesn't stop that, and instead just restricts the actual permission recalculation? am I missing something?

**Comment by perseus5** [comment]
> Sorry for the immense delay on this,
> I'm confused about the total intent of this, the PR states that it's preventing the command structure from being recalculated, but, it doesn't stop that, and instead just restricts the actual permission recalculation? am I missing something?

Yep, you're correct. The if statement needed to be moved down. I have since adjusted it and have tested it with @Plutoren (who was suffering from this issue) to make sure:
1) No more lag spikes on world changes
2) The bug that left you in an unpleasant state didn't show up again with these changes.

I believe from our testing both of these criteria have been met.

However, if anyone would like to test themselves, here's the jar: 
https://drive.google.com/file/d/1gf0yAUtvVC5FPmJh_NVXQUdBWsXW4Nu7/view?usp=sharing

**Comment by Plutoren** [comment]
Just confirming this. My very consistent 1 second lag spikes on world change have been completely removed after trying the most recent jar. I’m testing this on a production server and have had no complaints about other bugs being introduced. Switching worlds is incredibly fast now. There is still the same lag on any server join to send the command map but I suppose that’s unavoidable, however this is a huge help so thank you.

**Review by aikar — COMMENTED** [review]
the PaperConfig is still named about permissions and not about command map.

Though would be cool if we could store results of permission checks when building the command map.

Id suggest Map<String, Boolean> on EntityPlayer where key is command name, and value is if they had permission or not.

and then when this code is invoked, after permissions are recalculated, test every command in the cached results map, and if the result is different than previous, resend command map.

This would be nicer and wouldn't require a config, and could maybe even not do a full rebuild, but instead just the different nodes?

Thoughts?

**Comment by perseus5** [comment]
Fixed the settings name, I like the idea and it sounds good but I don't think I'd be able to implement such a system at this time.

**Pull request closed without merging**